// app/SlavicoinICO.js
/* eslint-disable global-require */
const express = require('express');
const logger = require('app/common/log/logger.service');
const bodyParser = require('body-parser');
const cookieParser = require('cookie-parser');
//const serveStatic = require('serve-static');
const glob = { glob: require('glob') };
const bluebird = require('bluebird');

global.Promise = bluebird;
Promise.promisifyAll(glob);

class SlavicoinICO {
  constructor() {

    this.connections = [];
    this.app = express();

    this.connectionClose = this.connectionClose.bind(this);
    this.connectionEstablished = this.connectionEstablished.bind(this);
    this.startServer = this.startServer.bind(this);
    this.stopServer = this.stopServer.bind(this);
    this.initializeServer = this.initializeServer.bind(this);
    this.includeAPIRoutes = this.includeAPIRoutes.bind(this);
  }

  initializeServer() {
    this.app.set('port', 3010);
    return this.includeAPIRoutes();
  }
  includeMiddlewares() {
    logger.info('Including middlewares');

    this.app.use('/api/*', bodyParser.urlencoded({ extended: true, limit: '20kb' }));
    this.app.use('/api/*', bodyParser.json({ limit: '20kb' }));

    this.app.use(cookieParser());
  }
  async includeAPIRoutes() {
    const routes = await glob.globAsync('**/api/**/*.route.js');
    routes.forEach((route) => {
      require(route.replace('node_modules/', ''))(this.app);
    });
  }
  connectionClose(closed) {
    this.connections = this.connections.filter(function (current) {
      return (current !== closed);
    });
  }
  connectionEstablished(connection) {
    this.connections.push(connection);
    connection.on('close', this.connectionClose);
  }
  startServer() {
    //logger.info('Starting server');
    this.server = this.app.listen(this.app.get('port'));

    this.server.on('connection', this.connectionEstablished);
  }
  stopServer(callback) {
    if(this.server) {
      this.server.close(callback);
    }
    this.connections.forEach(function (connection) {
      connection.end();
      connection.destroy();
    });
    if(!this.server) {
      return callback();
    }
  }
}

module.exports = SlavicoinICO;
